Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp

```diff
@@ -73,74 +93,203 @@ namespace AZ::DocumentPropertyEditor
             {
-                m_container->ClearElements(m_instance, impl->m_serializeContext);
+                m_container->ClearElements(m_containerInstance, impl->m_serializeContext);
 
                 auto containerNode = GetContainerNode(impl, path);
-                for (AZ::s64 i = m_container->Size(m_instance) - 1; i >= 0; --i)
-                {
-                    Nodes::PropertyEditor::RemoveNotify.InvokeOnDomNode(containerNode, i);
-                }
-                Nodes::PropertyEditor::ClearNotify.InvokeOnDomNode(containerNode);
                 Nodes::PropertyEditor::ChangeNotify.InvokeOnDomNode(containerNode);
-
                 impl->m_adapter->NotifyResetDocument();
             }
 
             void OnAddElement(ReflectionAdapterReflectionImpl* impl, const AZ::Dom::Path& path)
             {
-                if (m_container->IsFixedCapacity() && m_container->Size(m_instance) >= m_container->Capacity(m_instance))
+                if (m_container->IsFixedCapacity() && m_container->Size(m_containerInstance) >= m_container->Capacity(m_containerInstance))
                 {
                     return;
                 }
 
                 const AZ::SerializeContext::ClassElement* containerClassElement =
                     m_container->GetElement(m_container->GetDefaultElementNameCrc());
 
-                // dataAddress is allocated instance of key type. Get the key type and send it with the dataAddress in the
-                // message, then skip the store element below until we get an AddContainerKey message back from the DPE UI
-                void* dataAddress = m_container->ReserveElement(m_instance, containerClassElement);
+                // The reserved element is an allocated instance of the IDataContainer's ValueType.
+                // In an associative container, this would be a pair.
+                m_reservedElementInstance = m_container->ReserveElement(m_containerInstance, containerClassElement);
 
                 auto associativeContainer = m_container->GetAssociativeContainerInterface();
                 if (associativeContainer)
                 {
                     auto keyTypeAttribute = containerClassElement->FindAttribute(AZ_CRC_CE("KeyType"));
                     if (keyTypeAttribute)
                     {
+                        // Get the key type and send it with the dataAddress in the message, then skip the store
+                        // element below until we get an AddContainerKey message back from the DPE UI
                         auto* keyTypeData = azdynamic_cast<const AZ::Edit::AttributeData<AZ::Uuid>*>(keyTypeAttribute);
                         if (keyTypeData)
                         {
-                            auto keyType = keyTypeData->Get(nullptr);
-                            DocumentAdapterPtr reflectionAdapter = AZStd::make_shared<ReflectionAdapter>(dataAddress, keyType);
+                            auto& keyType = keyTypeData->Get(nullptr);
```

Comment:
This seems like it should be a `const auto&` or just a `const AZ::TypeId&`
```suggestion
                            const auto& keyType = keyTypeData->Get(nullptr);
```

Reply:
Done as well. Went with `const AZ::TypeId&`